### PR TITLE
Emit more assumptions in trans

### PR DESCRIPTION
**Pull request by arielb1**
Perf numbers pending.

### CODE REVIEW

File: src/librustc_trans/mir/lvalue.rs [hunk 1]
```diff
@@ -69,18 +67,6 @@ impl<'tcx> LvalueRef<'tcx> {
...
-}
-
-pub fn get_dataptr(b: &Builder, fat_ptr: ValueRef) -> ValueRef {
-    b.struct_gep(fat_ptr, abi::FAT_PTR_ADDR)
-}
-
-pub fn load_fat_ptr(b: &Builder, fat_ptr: ValueRef) -> (ValueRef, ValueRef) {
-    (b.load(get_dataptr(b, fat_ptr)), b.load(get_meta(b, fat_ptr)))
-}
-
```
Comment (eddyb) [inline src/librustc_trans/mir/lvalue.rs]: I prefer moving things to MIR, not away from it, TBH.

**Review by eddyb — COMMENTED** [review]
Only one complaint, everything else LGTM. r=me after changing that, if you want to.

File: src/librustc_trans/mir/lvalue.rs (hunk 1 above)
Comment (eddyb) [inline src/librustc_trans/mir/lvalue.rs]: Alright, we'll move them when we get rid of all non-MIR code.

Comment (arielb1) [inline src/librustc_trans/mir/lvalue.rs]: TBH, I would prefer that everything will be done through `OperandRef` and be in 1 file. That would require some more refactoring through.

**Comment by luqmana** [comment]
So, I recall we held back on adding assume to a lot of places cause it slowed down LLVM a bit. I'm not sure how much more efficient that's become?

cc @dotdash

**Comment by arielb1** [comment]
C-enum-to-integer casts are not really a common codepath - through we should see that this doesn't hit some terrible O(n^2) case with serialization or something.

File: src/librustc_trans/mir/lvalue.rs (hunk 1 above)
Comment (arielb1) [inline src/librustc_trans/mir/lvalue.rs]: I like the current structure where rvalue/block/stmt/lvalue are rather independent of each-other, and all "utility" methods go into a different place.

**Pull request merged**